enhancements/installer/custom-dns.md

```diff
@@ -0,0 +1,577 @@
+
+One of the decisions that needs to be made early in the design process is to
+determine the best way to order OpenShift's configuration of the Load Balancer
+and the customer's configuration on their custom DNS solution.
+
+Today, the Installer configures the LB first and the generated DNS name is used
+to configure DNS CNAME records. The LB DNS name cannot be predicted in advance
+because it is generated using an unique UUID while configuring LB entries for
+API and Ingress services.
+
+Here are a few options being considered as potential solutions:
+
+#### Option 1: Add a pre-install step to configure LB and provide DNS config
+
+This solution considers adding a pre-install step that configures Load Balancers
+for API and Ingress services. The DNS names generated per subnet/Availability Zone
+for API and Ingress services are output to the user in the form of DNS CNAME
+config. Customers can use this generated output to configure their DNS servers.
+
+This step could be run as an installer command like
+"openshift-install generate-custom-dns-config" while providing it with a complete
+install-config. After thsi step, when "create cluster" is called with the
+customDNS feature enabled, the validation step will have to ensure both the LB and
+DNS have been configured.
+
+An advantage to this approach is that at the end of "create cluster" the customer
+can expect to have a fully functioning cluster.
+
+This approach suggests a new way to get to a fully functional cluster in 2 parts.
+Does this constitute an "agent in the middle" install? The cluster install is not
+actively waiting for the DNS config to be completed by the customer. Another
+con is that a customer could run the "generate-custom-dns-config" utility
+but never follow through with the DNS config. This will result in dangling LB config.
+
+#### Option 1A: Generate LB and DNS config during pre-install step
+
+Another variation of this approach would be for this pre-install utility to generate
+LB DNS names but not actually configure the Load Balancers. One of the components
+of the LB DNS Name is the infra-id that is generated by the installer. If this
+pre-install utility is run to generate an infra-id, we should make sure to add it
+to the install-config so that eventual call to "create cluster" can use the same
+infra-id. 
+
+The current understanding is that the LB DNS name follows this pattern:
+<infra-id>-ext-<unknown-id>.elb.<region>.<cluster domanin name>
+
+[TODO: We are still figuring out if the "unknown id" is a randomly generated
+load-balancer-resource-id or a known subnet-id.]
+
+So, if the LB DNS name can be predicted by the installer (with the infra id being
+the only randomly generated value within the Installer), then the actual
+configuration of the LB can be postponed to the "create cluster" step with the
+pre-install step just generating the LB DNS name to be used for the DNS configuration.
+
+#### Option 2: Customer configures DNS as a post-install step
```

Comment:
I believe assisted takes a similar approach to this - it ensures that nothing it needs to do during installation relies on having the DNS set up (mostly by adding entries to /etc/hosts I think), and allows you to add it post-install. I don't know to what extent that is relying on the fact that it installs to platforms (baremetal/vsphere) that already run keepalived to provide an API VIP.
In principle I think this is the best user experience.

Reply:
Current understanding is that this is the best approach for this problem.